File: 잡동사니/k2022/spring/연습/board/src/main/java/com/karma/board/config/JpaConfig.java
package com.karma.board.config;

import com.karma.board.domain.MyPrincipal;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.domain.AuditorAware;
import org.springframework.data.jpa.repository.config.EnableJpaAuditing;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;

import java.util.Optional;

@EnableJpaAuditing
@Configuration
public class JpaConfig {
    @Bean
    public AuditorAware<String> auditorAware() {
        return () -> Optional
                // ⅰ) Security Context 가져오기
                .ofNullable(SecurityContextHolder.getContext())
                // ⅱ) 인증정보 꺼내기
                .map(SecurityContext::getAuthentication)
                // ⅲ) 인증여부 확인
                .filter(Authentication::isAuthenticated)
                // ⅳ) Principal 꺼내기
                .map(Authentication::getPrincipal)
                // ⅴ) Type Casting
                .map(MyPrincipal.class::cast)
                // ⅵ) 유저명 꺼내기
                .map(MyPrincipal::getUsername);
    }
}

File: backend/backend/src/main/kotlin/com/karma/chat/config/JpaConfig.kt
package com.karma.chat.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.data.domain.AuditorAware
import org.springframework.data.jpa.repository.config.EnableJpaAuditing
import org.springframework.security.core.context.SecurityContextHolder
import java.util.*

@EnableJpaAuditing
@Configuration
class JpaConfig {
    @Bean
    fun auditAware(): AuditorAware<String> = AuditorAware<String> {
        Optional.ofNullable(SecurityContextHolder.getContext())
            .map { it.authentication }
            .filter { it !=null && it.isAuthenticated }
            .map { it.name }
    }
}